Implement Richamster API library

shurikv opened the pull request:
Library for Richamster REST API

Reviewers: @Barsik-sus, @Sakapoi, @TheBestTvarynka, @n-romaniv

shurikv committed:
Implement Richamster API library

shurikv committed:
Update README.md

shurikv committed:
Update README.md

vsmysle commented on example/src/main.rs:
```diff
@@ -0,0 +1,149 @@
+    println!("username: {}", result.username);
+    println!("email: {}", result.email);
+    println!("role: {:?}", result.role);
+    println!("fee: {:?}", result.fee);
```
I think it can be beneficial to implement `std::fmt::Display` trait for a `User`. It will allow you to achieve the same with less code and enable users of your library to print `User` without printing every field.

shurikv replied:
I'm agree with you that better to impl basic display with output all fields in readable format

vsmysle commented on example/src/menu.rs:
```diff
@@ -0,0 +1,63 @@
+    UserDetails,
+    UserBalance,
+    Quit,
+}
```
Rust supports integer representation of the enums. 
```rust
#[repr(u8)]
pub enum MenuItems {
    Login = 1,
    UserDetails = 2,
    UserBalance = 3,
    Quit = 4
}
```
This will remove the necessity of implementing `impl From<MenuItems> for usize`.

shurikv replied:
Thanks! I never used this. Corrected

vsmysle commented on src/api/token.rs:
```diff
@@ -0,0 +1,219 @@
+            Token::LTC => "LTC",
+        }
+    }
+}
```
There is a `strum` crate which can do this for you via macro.
[Here](https://docs.rs/strum/latest/strum/derive.AsRefStr.html) you can find more details.

shurikv replied:
Thanks for advice. Used it

vsmysle commented on src/models/common.rs:
```diff
@@ -0,0 +1,104 @@
+{
+    let timestamp: String = Deserialize::deserialize(deserializer)?;
+    let datetime = Utc
+        .datetime_from_str(&timestamp, "%s")
```
Clippy shows that this method is deprecated and `DateTime::parse_from_str` should be used instead of it.

shurikv replied:
Corrected

vsmysle commented on src/models/common.rs:
```diff
@@ -0,0 +1,104 @@
+        return Ok(None);
+    }
+    let datetime = Utc
+        .datetime_from_str(&timestamp.unwrap(), "%s")
```
Clippy shows that this method is deprecated and DateTime::parse_from_str should be used instead of it.

shurikv replied:
Corrected

vsmysle commented on src/richamster/auth.rs:
```diff
@@ -0,0 +1,130 @@
+        match resp.status() {
+            StatusCode::OK => Ok(RequiresTwoFactor(true)),
+            StatusCode::CREATED => {
+                let token: TokenData = serde_json::from_str(&resp.text().await?)?;
+                Ok(Jwt(token.jwt_token))
```
Any particular reason why you are not using `response.json().await?` in your library?
Crate `reqwest` supports json response deserialization if you enable `json` feature.

shurikv replied:
Only that I only found out about it yesterday. Maybe in the future I will replace it

vsmysle commented on src/richamster/common.rs:
```diff
@@ -0,0 +1,101 @@
+use hmac_sha256::HMAC;
+use reqwest::RequestBuilder;
+use secrecy::{ExposeSecret, Secret};
```
It is pretty cool to see that you are using this crate for storing secret values🔥

shurikv replied:
Thanks

vsmysle commented on src/richamster/common.rs:
```diff
@@ -0,0 +1,101 @@
+    }
+}
+
+#[derive(Clone, Debug)]
+pub enum AuthState {
+    Unauthorized,
+    JwtTokenAuth(JwtToken),
+    ApiSecretKeyAuth(ApiKey, SecretKey),
+    JwtTokenWithApiSecretKeyAuth(JwtToken, ApiKey, SecretKey),
+}
```
I saw that you are using `AuthState::Unauthorized` as default state in some places (e.g `User`). If I understand the code correctly, you can derive `Default` implementation here and set `#[default]` for `Unauthorised`.

shurikv replied:
Done

vsmysle reviewed (approved):
Awesome work! Well done 👍

shurikv committed:
Refactoring code. Add transanctions list to CLI example

shurikv committed:
Impl Display for UserDetail struct

Dzuchun commented on src/models/user.rs:
```diff
@@ -0,0 +1,201 @@
+        if let Some(closed_at_lte) = &self.closed_at_lte {
+            result.push(format!("closed_at__lte={}", closed_at_lte));
+        }
+        result.join("&")
```
Would like to remind, that there's a direct way to add separate parameters to the query with [`Url::query_pairs_mut`](https://docs.rs/url/latest/url/struct.Url.html#method.query_pairs_mut).

I guess, It's not a big deal, but would probably make code more readable.

shurikv replied:
Good point, thanks. Rewrote code

Dzuchun reviewed (commented):
:+1:

shurikv committed:
Refactoring code

- Use reqwest json feature to serialize/deserialize message
- Use methods from Url crate to add parameters to url
- Add reqwest::Client to User, Exchange, Payments structs and use it to send requests

shurikv committed:
Refactoring code

- Improve macros for send request
- Remove function and use macros for send requests
- Remove Client from User, Exchange, Payments structs
- Simplify code

shurikv committed:
Add Hash and Eq derive for Token and CurrencyPair to use it in HashMap. Correct visibility fields in structs
